File: frontend/apps/ios/Pantopus/Features/Homes/MyHomesListViewModel.swift
//
//  MyHomesListViewModel.swift
//  Pantopus
//
//  Backs `MyHomesListView`. Fetches `GET /api/homes/my-homes` and
//  projects each home into the avatar-first List-of-Rows row shape
//  defined for T6.3f / P14:
//
//    • Leading — identity-green avatar tile (initials from address)
//    • Title   — nickname or formatted address
//    • Subtitle — role chip + locality (joined via "·")
//    • Body    — "Active home" home-tinted chip on the primary-owner row
//    • Trailing — chevron (tap → home dashboard)
//
//  Plus a `BannerConfig` intro card ("homes you belong to") and a 60pt
//  secondary-create FAB tinted home green.
//

import Foundation
import Observation
import SwiftUI

/// ViewModel for the "My homes" list.
@Observable
@MainActor
final class MyHomesListViewModel: ListOfRowsDataSource {
    let title = "My homes"
    var topBarAction: TopBarAction? {
        nil
    }

    let tabs: [ListOfRowsTab] = []
    var selectedTab: String = ""
    var fab: FABAction? {
        FABAction(
            icon: .plusCircle,
            accessibilityLabel: "Claim a home",
            variant: .secondaryCreate,
            tint: .home,
            handler: onAddHome
        )
    }

    var banner: BannerConfig? {
        guard case let .loaded(sections, _) = state,
              let count = sections.first?.rows.count,
              count > 0
        else {
            return nil
        }
        return BannerConfig(
            icon: .home,
            title: count == 1 ? "1 home you belong to" : "\(count) homes you belong to",
            subtitle: "Tap any home to jump into that household",
            tint: .home
        )
    }

    private(set) var state: ListOfRowsState = .loading

    private let api: APIClient
    private let onOpenHome: (String) -> Void
    private let onAddHome: @Sendable () -> Void

    init(
        api: APIClient = .shared,
        onOpenHome: @escaping (String) -> Void = { _ in },
        onAddHome: @escaping @Sendable () -> Void = {}
    ) {
        self.api = api
        self.onOpenHome = onOpenHome
        self.onAddHome = onAddHome
    }

    func load() async {
        if case .loaded = state { return }
        state = .loading
        await fetch()
    }

    func refresh() async {
        await fetch()
    }

    func loadMoreIfNeeded() async {} // `my-homes` is not paginated.

    private func fetch() async {
        do {
            let response: MyHomesResponse = try await api.request(HomesEndpoints.myHomes())
            let rows = response.homes.map { row(for: $0) }
            if rows.isEmpty {
                state = .empty(
                    ListOfRowsState.EmptyContent(
                        icon: .home,
                        headline: "You don’t belong to any homes yet",
                        subcopy: "Claim or join a verified home to unlock packages, bills, tasks, and member chat.",
                        ctaTitle: "Claim a home",
                        onCTA: onAddHome
                    )
                )
            } else {
                state = .loaded(sections: [RowSection(rows: rows)], hasMore: false)
            }
        } catch {
            state = .error(message: (error as? APIError)?.errorDescription ?? "Something went wrong.")
        }
    }

    private func row(for entry: MyHome) -> RowModel {
        let home = entry.home
        let displayTitle = home.name?.nilIfEmpty
            ?? home.address?.nilIfEmpty
            ?? "Unnamed home"
        let locality = [home.city, home.state]
            .compactMap { $0?.nilIfEmpty }
            .joined(separator: ", ")
            .nilIfEmpty
        let role = roleLabel(for: entry)
        let subtitleParts = [role, locality].compactMap { $0 }
        let subtitle = subtitleParts.isEmpty ? nil : subtitleParts.joined(separator: " · ")

        let chips: [RowChip]? = if entry.isPrimaryOwner == true {
            [
                RowChip(
                    text: "Active home",
                    icon: .home,
                    tint: .custom(
                        background: Theme.Color.homeBg,
                        foreground: Theme.Color.home
                    )
                )
            ]
        } else {
            nil
        }

        return RowModel(
            id: entry.id,
            title: displayTitle,
            subtitle: subtitle,
            template: .avatarKebab,
            leading: .avatar(
                name: displayTitle,
                imageURL: nil,
                identity: .home,
                ringProgress: entry.ownershipStatus == "verified" ? 1.0 : 0.3
            ),
            trailing: .chevron,
            onTap: { @Sendable in
                Task { @MainActor in self.onOpenHome(entry.id) }
            },
            chips: chips
        )
    }

    /// Maps the backend's role enum onto the canonical four-role label
    /// vocabulary the design uses: Owner / Tenant / Housemate / Guest.
    /// Order: ownership_status wins; otherwise occupancy.role_base; final
    /// fallback "Member".
    private func roleLabel(for entry: MyHome) -> String? {
        if let status = entry.ownershipStatus {
            switch status {
            case "verified":
                return "Owner"
            case "pending":
                return "Owner (pending)"
            default:
                break
            }
        }
        switch entry.occupancy?.roleBase {
        case "lease_resident":
            return "Tenant"
        case "household_member":
            return "Housemate"
        case "guest":
            return "Guest"
        case "owner":
            return "Owner"
        case "admin", "manager":
            return "Manager"
        case nil:
            return nil
        case let roleBase?:
            return roleBase.capitalized
        }
    }
}

private extension String {
    /// Returns `nil` when the string is empty; otherwise the string.
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/MyHomesListViewModel.kt
@file:Suppress("MagicNumber")

package app.pantopus.android.ui.screens.homes

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.homes.MyHome
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.homes.HomesRepository
import app.pantopus.android.ui.components.IdentityPillar
import app.pantopus.android.ui.screens.shared.list_of_rows.BannerConfig
import app.pantopus.android.ui.screens.shared.list_of_rows.BannerCtaTint
import app.pantopus.android.ui.screens.shared.list_of_rows.ListOfRowsUiState
import app.pantopus.android.ui.screens.shared.list_of_rows.RowChip
import app.pantopus.android.ui.screens.shared.list_of_rows.RowLeading
import app.pantopus.android.ui.screens.shared.list_of_rows.RowModel
import app.pantopus.android.ui.screens.shared.list_of_rows.RowSection
import app.pantopus.android.ui.screens.shared.list_of_rows.RowTemplate
import app.pantopus.android.ui.screens.shared.list_of_rows.RowTrailing
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for the refreshed My homes list — wraps `GET /api/homes/my-homes`.
 *
 * T6.3f / P14 row anatomy:
 *   leading  → identity-green avatar tile (initials from address)
 *   title    → nickname or formatted address
 *   subtitle → role chip + locality joined with "·"
 *   chips    → ["Active home"] on the primary-owner row (home-tinted)
 *   trailing → chevron (tap → Home dashboard)
 *
 * Plus a home-tinted intro banner ("N homes you belong to") and a
 * `.SecondaryCreate` FAB tinted `FabTint.Home`.
 */
@HiltViewModel
class MyHomesListViewModel
    @Inject
    constructor(
        private val repo: HomesRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow<ListOfRowsUiState>(ListOfRowsUiState.Loading)
        val state: StateFlow<ListOfRowsUiState> = _state.asStateFlow()

        private val _banner = MutableStateFlow<BannerConfig?>(null)
        val banner: StateFlow<BannerConfig?> = _banner.asStateFlow()

        private var onOpenHome: (String) -> Unit = {}
        private var onAddHome: () -> Unit = {}

        fun configureNavigation(
            onOpenHome: (String) -> Unit,
            onAddHome: () -> Unit,
        ) {
            this.onOpenHome = onOpenHome
            this.onAddHome = onAddHome
        }

        fun load() {
            if (_state.value is ListOfRowsUiState.Loaded) return
            refresh()
        }

        fun refresh() {
            _state.value = ListOfRowsUiState.Loading
            _banner.value = null
            viewModelScope.launch {
                when (val result = repo.myHomes()) {
                    is NetworkResult.Success -> applySuccess(result.data.homes)
                    is NetworkResult.Failure -> _state.value = ListOfRowsUiState.Error(result.error.message)
                }
            }
        }

        private fun applySuccess(homes: List<MyHome>) {
            if (homes.isEmpty()) {
                _state.value =
                    ListOfRowsUiState.Empty(
                        icon = PantopusIcon.Home,
                        headline = "You don’t belong to any homes yet",
                        subcopy = "Claim or join a verified home to unlock packages, bills, tasks, and member chat.",
                        ctaTitle = "Claim a home",
                        onCta = onAddHome,
                    )
                return
            }
            val rows = homes.map(::rowFor)
            _state.value =
                ListOfRowsUiState.Loaded(
                    sections = listOf(RowSection(id = "my-homes", rows = rows)),
                    hasMore = false,
                )
            _banner.value =
                BannerConfig(
                    icon = PantopusIcon.Home,
                    title = if (rows.size == 1) "1 home you belong to" else "${rows.size} homes you belong to",
                    subtitle = "Tap any home to jump into that household",
                    tint = BannerCtaTint.Home,
                )
        }

        private fun rowFor(home: MyHome): RowModel {
            val title =
                home.name?.takeIf { it.isNotEmpty() }
                    ?: home.address
                    ?: "Unnamed home"
            val locality =
                listOfNotNull(home.city, home.state)
                    .filter { it.isNotEmpty() }
                    .joinToString(", ")
                    .takeIf { it.isNotEmpty() }
            val role = roleLabel(home)
            val subtitle =
                listOfNotNull(role, locality)
                    .joinToString(" · ")
                    .takeIf { it.isNotEmpty() }
            val progress = if (home.ownershipStatus == "verified") 1.0f else 0.3f

            val chips: List<RowChip>? =
                if (home.isPrimaryOwner == true) {
                    listOf(
                        RowChip(
                            text = "Active home",
                            icon = PantopusIcon.Home,
                            tint =
                                RowChip.Tint.Custom(
                                    background = PantopusColors.homeBg,
                                    foreground = PantopusColors.home,
                                ),
                        ),
                    )
                } else {
                    null
                }

            return RowModel(
                id = home.id,
                title = title,
                subtitle = subtitle,
                template = RowTemplate.AvatarKebab,
                leading =
                    RowLeading.Avatar(
                        name = title,
                        imageUrl = null,
                        identity = IdentityPillar.Home,
                        ringProgress = progress,
                    ),
                trailing = RowTrailing.Chevron,
                onTap = { onOpenHome(home.id) },
                chips = chips,
            )
        }

        /**
         * Maps the backend's role hierarchy onto the canonical four-role
         * label vocabulary the design uses: Owner / Tenant / Housemate /
         * Guest. `ownership_status` wins; otherwise `occupancy.role_base`;
         * final fallback `null` so the subtitle just shows locality.
         */
        private fun roleLabel(home: MyHome): String? {
            when (home.ownershipStatus) {
                "verified" -> return "Owner"
                "pending" -> return "Owner (pending)"
                else -> Unit
            }
            return when (home.occupancy?.roleBase) {
                "lease_resident" -> "Tenant"
                "household_member" -> "Housemate"
                "guest" -> "Guest"
                "owner" -> "Owner"
                "admin", "manager" -> "Manager"
                null -> null
                else -> home.occupancy?.roleBase?.replaceFirstChar { it.uppercase() }
            }
        }
    }
